pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/CassandraSinkTask.java

```diff
@@ -189,28 +199,22 @@ public void write(Record<T> record) throws Exception {
         log.info("raw {}", rawvalue);
       }
     }
-    int number;
-    synchronized (this) {
-      incomingList.add(record);
-      number = incomingList.size();
-    }
-    if (number == batchSize) {
+
+    incomingQueue.put(record); // blocking call to slow down the topic consumer
+    if (incomingQueue.size() >= batchSize) {
       flushExecutor.submit(this::flush);
     }
   }
 
   protected void flush() {
-    final List<Record<T>> swapList;
-    synchronized (this) {
```

Comment:
Not needed anymore because the atomic `isFlushing` guarantees no two flush tasks run at the same time. Also the queue operations are thread safe and capped at `batchSize`